frontend/lib/resolution/scope-queries.cpp

```diff
@@ -2196,17 +2196,136 @@ std::vector<ID> findUsedImportedModules(Context* context,
   auto result = resolveVisibilityStmts(context, scope);
   std::vector<ID> ids;
 
+  if (result == nullptr) return ids;
+
   for (const auto& r : result->visibilityClauses()) {
     ids.push_back(r.scope()->id());
   }
   return ids;
 }
 
-void resolveUsesAndImportsInScope(Context* context, const Scope* scope) {
-  std::ignore = resolveVisibilityStmts(context, scope);
-  return;
+static const std::map<ID, ID>&
+findAllModulesUsedImportedInTreeQuery(Context* context, ID id);
+
+// Pre-order depth first traveral of the entire module to gather use/import.
+// Key is ID of use/import, value is target module ID. We use a map here
+// because the C++ standard map maintains ordering, which should sort all
+// the use/import by their lexical order (via ID ordering).
+static std::map<ID, ID>
+findAllModulesUsedImportedInTreeImpl(Context* context, ID id) {
+  if (id.isEmpty()) return {};
+
+  auto ast = parsing::idToAst(context, id);
+  std::map<ID, ID> ret;
+
+  if (auto scope = scopeForId(context, id)) {
+    if (auto rvs = resolveVisibilityStmts(context, scope)) {
+      for (auto& vc : rvs->visibilityClauses()) {
+        auto& k = vc.visibilityClauseId();
+        auto& v = vc.scope()->id();
+        ret.insert(std::make_pair(k, v));
+      }
+    }
+  }
+
+  // Recurse, but only use a query if the child is a module.
```

Comment:
I would think it would not recurse if the child is a module. Does the spec traversal order preclude that?

Reply:
I think you're right, and while children need to visit parents, the reverse may not be true.